oxenc/endian.h

```diff
@@ -4,26 +4,64 @@
 #include <concepts>
 #include <cstdint>
 #include <cstring>
+#include <memory>
+#include <utility>
 
 #if defined(_MSC_VER) && (!defined(__clang__) || defined(__c2__))
 #include <cstdlib>
 
+#elif !(defined(__clang__) || defined(__GNUC__))
+#if defined(__linux__)
+extern "C" {
+#include <byteswap.h>
+}  // extern "C"
+
+#else
+#include <algorithm>
+
+#endif
+#endif
+
+namespace oxenc {
+#ifdef __cpp_lib_bit_cast
+using std::bit_cast;
+#else
+template <class To, class From>
+    requires(
+            sizeof(To) == sizeof(From) && std::is_trivially_copyable<To>::value &&
+            std::is_trivially_copyable<From>::value && std::is_trivially_constructible<To>::value &&
+            std::is_trivially_constructible<From>::value)
+inline constexpr To bit_cast(const From& from) {
+    To storage{};
+    std::construct_at(std::launder(&storage), from);
+    return storage;
+}
+#endif
+
+#if defined(_MSC_VER) && (!defined(__clang__) || defined(__c2__))
 #define bswap_16(x) _byteswap_ushort(x)
 #define bswap_32(x) _byteswap_ulong(x)
 #define bswap_64(x) _byteswap_uint64(x)
+
 #elif defined(__clang__) || defined(__GNUC__)
 #define bswap_16(x) __builtin_bswap16(x)
 #define bswap_32(x) __builtin_bswap32(x)
 #define bswap_64(x) __builtin_bswap64(x)
-#elif defined(__linux__)
-extern "C" {
-#include <byteswap.h>
-}  // extern "C"
-#else
-#error "Don't know how to byteswap on this platform!"
-#endif
 
-namespace oxenc {
+#elif !defined(__linux__)
+template <std::integral T>
+    requires std::has_unique_object_representations_v<T>
+inline constexpr T byteswap(T val) noexcept {
+    auto value = bit_cast<std::array<std::byte, sizeof(T)>>(val);
+    std::ranges::reverse(value);
+    return bit_cast<T>(value);
+}
```

Comment:
I'm a bit concerned that a compiler isn't going to generate optimal *run-time* code here, and so we end up adding constexpr support at the cost of runtime performance.

But I think we can avoid the potential runtime penalty by keeping the range reverse algorithm for constexpr evaluation, but going back to the intrinsics for runtime evaluation via `std::is_constant_evaluated()`:

```C++
if (std::is_constant_evaluated()) {
    // above implementation
} else {
   // call to potentially non-constexpr intrinsic
}
```

Reply:
Following up on off-band discussion: updated platform-specific swapping and added an extra commit with a speed test for comparison